Improve series matching

fredriklindberg opened the pull request:
Improves series name matching (thus better hit ratio at tmdb) for paths that contain more information that just the series name (ex. show.s01). The output from the SeriesResolver is a "clean" series name.

crobibero commented on Emby.Server.Implementations/Library/LibraryManager.cs:
```diff
@@ -2688,6 +2688,21 @@ public NamingOptions GetNamingOptions()
+
+            return new ItemLookupInfo
+            {
+                Name = episodeInfo.SeriesName.Replace(".", " ", true, CultureInfo.InvariantCulture)
```
What happens with a series such as [S.W.A.T.](https://www.tvmaze.com/shows/21532/swat)?

crobibero commented on MediaBrowser.Providers/Plugins/Tmdb/TV/TmdbSeriesProvider.cs:
```diff
@@ -186,9 +186,19 @@ public async Task<MetadataResult<Series>> GetMetadata(SeriesInfo info, Cancellat
+                if (seriesName != null)
+                {
+                    parsedName = seriesName;
+                } else
+                {
```
```suggestion
                }
                else
                {
```

crobibero commented on Emby.Naming/Common/NamingOptions.cs:
```diff
@@ -137,7 +137,7 @@ public NamingOptions()
             CleanStrings = new[]
             {
-                @"[ _\,\.\(\)\[\]\-](3d|sbs|tab|hsbs|htab|mvc|HDR|HDC|UHD|UltraHD|4k|ac3|dts|custom|dc|divx|divx5|dsr|dsrip|dutch|dvd|dvdrip|dvdscr|dvdscreener|screener|dvdivx|cam|fragment|fs|hdtv|hdrip|hdtvrip|internal|limited|multisubs|ntsc|ogg|ogm|pal|pdtv|proper|repack|rerip|retail|cd[1-9]|r3|r5|bd5|bd|se|svcd|swedish|german|read.nfo|nfofix|unrated|ws|telesync|ts|telecine|tc|brrip|bdrip|480p| ...
+                @"[ _\,\.\(\)\[\]\-](([sS][0-9]{1,3}([eE][0-9]+){0,1})|([0-9]{1,3}[xX]([0-9]+){0,1})|3d|sbs|tab|hsbs|htab|mvc|HDR|HDC|UHD|UltraHD|4k|ac3|dts|custom|dc|divx|divx5|dsr|dsrip|dutch|dvd|dvdrip|dvdscr|dvdscreener|screener|dvdivx|cam|fragment|fs|hdtv|hdrip|hdtvrip|internal|limited|multisubs|ntsc|ogg|ogm|pal|pdtv|proper|repack|rerip|retail|cd[1-9]|r3|r5|bd5|bd|se|svcd|swedish|german|read ...
```
Since you've edited the Regex, can you please add tests to support the new naming format?

fredriklindberg commented:
I've addressed the PR comments, also re-worked the cleanstring regex further and added more tests.

This change significantly improves match rate for a new database with entries without a prior tmdb id, especially after the deprecation of tvdb.

cvium commented on Emby.Naming/Common/NamingOptions.cs:
```diff
@@ -137,7 +137,7 @@ public NamingOptions()
             CleanStrings = new[]
             {
-                @"[ _\,\.\(\)\[\]\-](3d|sbs|tab|hsbs|htab|mvc|HDR|HDC|UHD|UltraHD|4k|ac3|dts|custom|dc|divx|divx5|dsr|dsrip|dutch|dvd|dvdrip|dvdscr|dvdscreener|screener|dvdivx|cam|fragment|fs|hdtv|hdrip|hdtvrip|internal|limited|multisubs|ntsc|ogg|ogm|pal|pdtv|proper|repack|rerip|retail|cd[1-9]|r3|r5|bd5|bd|se|svcd|swedish|german|read.nfo|nfofix|unrated|ws|telesync|ts|telecine|tc|brrip|bdrip|480p| ...
+                @"([ \.\,_\-]*([sS][0-9]+([eE][0-9]+)*|[0-9]+[xX][0-9]+)([ _\,\.\(\)\[\]\-][^ _\,\.\(\)\[\]\-]+)+)|([ \.\,_\-]+(\[.*\]|3d|sbs|tab|hsbs|htab|mvc|HDR|HDC|UHD|UltraHD|4k|ac3|dts|custom|dc|divx|divx5|dsr|dsrip|dutch|dvd|dvdrip|dvdscr|dvdscreener|screener|dvdivx|cam|fragment|fs|hdtv|hdrip|hdtvrip|internal|limited|multisubs|ntsc|ogg|ogm|pal|pdtv|proper|repack|rerip|retail|cd[1-9]|r3|r5| ...
```
I don't like having series related stuff in this regex tbh. It is supposed to be a generic string cleaning regex.

fredriklindberg replied:
Fair enough

cvium commented on MediaBrowser.Controller/Library/ILibraryManager.cs:
```diff
@@ -410,6 +410,13 @@ public interface ILibraryManager
+        /// </summary>
+        /// <param name="path">The path.</param>
+        /// <returns>ItemInfo.</returns>
+        ItemLookupInfo GetSeriesFromPath(string path);
```
I know that we already have `GetSeasonNumberFromPath` in this interface, but I do wonder if it would not be better suited for a new `SeriesResolver`. It's pretty terrible design to have media type specific methods in the library manager imo.

Or maybe using EpisodePathParser directly in tmdb is better?

Or better yet, extending `EpisodeInfo` with the SeriesName so it's available to all providers automatically? `EpisodeInfo` is populated here https://github.com/jellyfin/jellyfin/blob/master/MediaBrowser.Controller/Entities/TV/Episode.cs#L305

fredriklindberg replied:
Yeah, I don't really like that the name have to be "cleaned" inside the provider. It should be fed a clean name, but it's not happening for the series provider right now. (it gets things like "foobar s01e1" which makes tmdb fail).

I'll see if I can figure out something better than the clean regex.

cvium replied:
At a glance I don't see any problems with extending `GetLookupInfo` for Series, Season and Episode to include the clean series name. And shouldn't it fix the issue as well (all metadata providers will have to be updated to use it of course)?

cvium commented on Emby.Server.Implementations/Library/LibraryManager.cs:
```diff
@@ -2699,6 +2699,21 @@ public NamingOptions GetNamingOptions()
 
+        public ItemLookupInfo GetSeriesFromPath(string path)
+        {
+            var episodeParser = new EpisodeResolver(new NamingOptions());
```
Don't create a new namingoptions. Use `GetNamingOptions()`.

cvium commented on Emby.Server.Implementations/Library/LibraryManager.cs:
```diff
@@ -2699,6 +2699,21 @@ public NamingOptions GetNamingOptions()
+        {
+            var episodeParser = new EpisodeResolver(new NamingOptions());
+            var episodeInfo = episodeParser.Resolve(path, true, true, false, false, true);
+            if (episodeInfo == null || string.IsNullOrEmpty(episodeInfo.SeriesName))
```
Could be shortened to `string.IsNullOrEmpty(episodeInfo?.SeriesName)`

cvium commented on Emby.Naming/TV/SeriesResolver.cs:
```diff
@@ -0,0 +1,64 @@
+    /// <summary>
+    /// Used to resolve information about series from path.
+    /// </summary>
+    public class SeriesResolver
```
This class is better suited as a static class. There's no state and only a single public method.

fredriklindberg replied:
Fixed

cvium commented on Emby.Naming/TV/SeriesPathParser.cs:
```diff
@@ -0,0 +1,95 @@
+    /// Used to parse information about series from paths containing more information that only the series name.
+    /// Uses the same regular expressions as the EpisodePathParser but have different success criteria.
+    /// </summary>
+    public class SeriesPathParser
```
I also think this should be static.

fredriklindberg replied:
Fixed

cvium commented on Emby.Naming/TV/SeriesInfo.cs:
```diff
@@ -0,0 +1,34 @@
+        /// <summary>
+        /// Gets or sets optional season number.
+        /// </summary>
+        public int? SeasonNumber { get; set; }
```
Why is this needed? Feels a bit out of place.

fredriklindberg replied:
Fixed, removed the season number

cvium commented on Emby.Naming/TV/SeriesPathParser.cs:
```diff
@@ -0,0 +1,95 @@
+
+                // Invalidate match when the season is 200 through 1927 or above 2500
+                // because it is an error unless the TV show is intentionally using false season numbers.
+                // It avoids erroneous parsing of something like "Series Special (1920x1080).mkv" as being season 1920 episode 1080.
```
Should have a test for this

fredriklindberg replied:
This code block is gone since I removed the season numbers.

fredriklindberg commented on Emby.Naming/TV/SeriesResolver.cs:
```diff
@@ -0,0 +1,64 @@
+            return new SeriesInfo(path)
+            {
+                Name = seriesName,
+                SeasonNumber = seasonNumber
```
Fixed, removed the season number

fredriklindberg commented on Emby.Naming/TV/SeriesResolver.cs:
```diff
@@ -0,0 +1,64 @@
+        /// <summary>
+        /// Regex that matches strings of at least 2 characters separated by a dot or underscore.
+        /// </summary>
+        private static readonly Regex _seriesNameRegex = new Regex(@"((?<a>[^\._]{2,})[\._]*)|([\._](?<b>[^\._]{2,}))");
```
Turns things like "The_Show or "The.Show" into "The Show". But still preserves names like "S.H.O.W" or "Show A.B". Added a comment to clarify.

cvium commented on Emby.Naming/TV/SeriesPathParser.cs:
```diff
@@ -0,0 +1,64 @@
+                    result.SeriesName = result.SeriesName
+                        .Trim()
+                        .Trim('_', '.', '-')
+                        .Trim();
```
Minor optimization thing (but given a few thousand items might add up), I wonder how often do we encounter special unicode whitespace chars? Would this `.Trim(' ', '_', '.', '-')` not cover something like 99% of cases?

Right now we're potentially creating 3 new strings for every file. The good thing about `.Trim()` is that it trims _all_ whitespace chars, but I can't help but wonder if it's even necessary.

fredriklindberg replied:
Probably good enough with  `.Trim(' ', '_', '.', '-')` yes

fredriklindberg replied:
Fixed

cvium commented on Emby.Naming/Common/NamingOptions.cs:
```diff
@@ -368,6 +368,20 @@ public NamingOptions()
+
+                // Series and season only expression
+                // "the show/season 1", "the show/s01"
+                new EpisodeExpression(@"(.*(\\|\/))*(?<seriesname>.+)\/[Ss](eason){0,1}[\. _\-]*(?<seasonnumber>\d+)")
```
We should always refrain from using `\d` as it covers a whole lot of unicode digits as well.
```suggestion
                new EpisodeExpression(@"(.*(\\|\/))*(?<seriesname>.+)\/[Ss](eason){0,1}[\. _\-]*(?<seasonnumber>[0-9]+)")
```

fredriklindberg replied:
Fixed

cvium commented on Emby.Naming/Common/NamingOptions.cs:
```diff
@@ -368,6 +368,20 @@ public NamingOptions()
+
+                // Series and season only expression
+                // "the show S01", "the show season 1"
+                new EpisodeExpression(@"(.*(\\|\/))*(?<seriesname>.+)[\. _\-]+[sS](eason){0,1}[\. _\-]*(?<seasonnumber>\d+)")
```
```suggestion
                new EpisodeExpression(@"(.*(\\|\/))*(?<seriesname>.+)[\. _\-]+[sS](eason){0,1}[\. _\-]*(?<seasonnumber>[0-9]+)")
```

fredriklindberg replied:
Fixed

cvium commented on Emby.Naming/TV/SeriesPathParser.cs:
```diff
@@ -0,0 +1,64 @@
+                if (expression.IsNamed)
+                {
+                    result.SeriesName = match.Groups["seriesname"].Value;
+                    result.Success = !string.IsNullOrEmpty(result.SeriesName) && !string.IsNullOrEmpty(match.Groups["seasonnumber"]?.Value);
```
Does it matter if seasonnumber is empty?

fredriklindberg replied:
The path parser tries to be a bit "smart" in figuring out what's a series name or not, it does that by looking at common naming patterns. If the path parser fails the series resolver will just default back to the path name (as how it was before). 

So if there is a series path without any season identifier, the path will be returned by the resolver as is. See the tests for the series resolver.

fredriklindberg replied:
On second thought, the condition could be changed to be (episode number or season identifier). But would need some test for that to make it work.

cvium replied:
Do you still want to do this?

fredriklindberg replied:
Do you have any sample data I can use for the test cases (I'm not familiar with anime)?

Bond-009 commented on Emby.Naming/Common/NamingOptions.cs:
```diff
@@ -368,6 +368,20 @@ public NamingOptions()
+
+                // Series and season only expression
+                // "the show/season 1", "the show/s01"
+                new EpisodeExpression(@"(.*(\\|\/))*(?<seriesname>.+)\/[Ss](eason){0,1}[\. _\-]*(?<seasonnumber>[0-9]+)")
+                {
+                    IsNamed = true
+                },
+
+                // Series and season only expression
+                // "the show S01", "the show season 1"
+                new EpisodeExpression(@"(.*(\\|\/))*(?<seriesname>.+)[\. _\-]+[sS](eason){0,1}[\. _\-]*(?<seasonnumber>[0-9]+)")
```
```suggestion
                new EpisodeExpression(@"(.*(\\|\/))*(?<seriesname>.+)\/[Ss](eason)?[\. _\-]*(?<seasonnumber>[0-9]+)")
                {
                    IsNamed = true
                },

                // Series and season only expression
                // "the show S01", "the show season 1"
                new EpisodeExpression(@"(.*(\\|\/))*(?<seriesname>.+)[\. _\-]+[sS](eason)?[\. _\-]*(?<seasonnumber>[0-9]+)")
```

fredriklindberg replied:
Fixed
